lib/srv/git/forward.go

```diff
@@ -0,0 +1,590 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(accessChecker.CheckAccess(s.cfg.TargetServer, state))
+}
+
+func (s *ForwardServer) onConnection(ctx context.Context, ccx *sshutils.ConnectionContext) (context.Context, error) {
+	s.logger.DebugContext(ctx, "On new connection")
+
+	identityCtx, err := s.auth.CreateIdentityContext(ccx.ServerConn)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.initRemoteConn(ctx, ccx, identityCtx); err != nil {
+		s.logger.DebugContext(ctx, "onConnection failed", "error", err)
+		return ctx, trace.Wrap(err)
+	}
+
+	// TODO(greedy52) decouple from srv.NewServerContext. We only need
+	// connection monitoring.
+	serverCtx, err := srv.NewServerContext(ctx, ccx, s, identityCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s.logger.DebugContext(ctx, "New connection accepted")
+	ccx.AddCloser(serverCtx)
+	return ctx, nil
+}
+
+func (s *ForwardServer) onChannel(ctx context.Context, ccx *sshutils.ConnectionContext, nch ssh.NewChannel) {
+	s.logger.DebugContext(ctx, "On new channel", "channel", nch.ChannelType())
+
+	// Only expecting a session to execute a command.
+	if nch.ChannelType() != teleport.ChanSession {
+		s.reply.RejectUnknownChannel(ctx, nch)
+		return
+	}
+
+	if s.remoteClient == nil {
+		s.reply.RejectWithNewRemoteSessionError(ctx, nch, trace.NotFound("missing remote client"))
+		return
+	}
+	remoteSession, err := s.remoteClient.NewSession(ctx)
+	if err != nil {
+		s.reply.RejectWithNewRemoteSessionError(ctx, nch, err)
+		return
+	}
+	defer remoteSession.Close()
+
+	ch, in, err := nch.Accept()
+	if err != nil {
+		s.reply.RejectWithAcceptError(ctx, nch, err)
+		return
+	}
+	defer ch.Close()
+
+	for {
+		select {
+		case req := <-in:
+			if req == nil {
+				s.logger.DebugContext(ctx, "Client disconnected", "remote_addr", ccx.ServerConn.RemoteAddr())
+				return
+			}
+
+			ok, err := s.dispatch(ctx, ch, req, remoteSession)
+			if err != nil {
+				s.reply.ReplyError(ctx, ch, req, err)
+				return
+			}
+			s.reply.ReplyRequest(ctx, req, ok, nil)
+
+		case execErr := <-s.waitExec:
+			code := sshutils.ExitCodeFromExecError(execErr)
+			s.logger.DebugContext(ctx, "Exec request complete", "code", code)
+			s.reply.SendExitStatus(ctx, ch, code)
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// dispatch executes an incoming request. If successful, it returns the ok value
+// for the reply. Otherwise, it returns the error it encountered.
+func (s *ForwardServer) dispatch(ctx context.Context, ch ssh.Channel, req *ssh.Request, remoteSession *tracessh.Session) (bool, error) {
+	s.logger.DebugContext(ctx, "Dispatching client request", "request_type", req.Type)
+
+	switch req.Type {
+	case tracessh.EnvsRequest:
+		s.logger.DebugContext(ctx, "Ignored request", "request_type", req.Type)
+		return true, nil
+	case sshutils.ExecRequest:
+		return true, trace.Wrap(s.handleExec(ctx, ch, req, remoteSession))
+	case sshutils.EnvRequest:
```

Comment:
Do we ever want and/or need envvars in the ssh git protocol?

Reply:
openssh sends `LANG` and `GIT_PROTOCOL`. it would be nice to forward them.